system: improve handling around GetTotalRAM()

vasild opened the pull request:
1. Fix unused variable warning (https://github.com/bitcoin/bitcoin/pull/33333#discussion_r2362493046)
2. Enable `GetTotalRAM()` on other platforms where it was tested to work.
3. Skip the `GetTotalRAM()` unit test on unsupported platforms.

Prior discussion: https://github.com/bitcoin/bitcoin/pull/33333#discussion_r2362493046

vasild committed:
system: improve handling around GetTotalRAM()

This patch achieves two things:
1. Fix unused variable warning (https://github.com/bitcoin/bitcoin/pull/33333#discussion_r2362493046)
2. Enable GetTotalRAM() on other platforms where it was tested to work.

Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

hebasto reviewed (commented):
Concept ACK.

hebasto commented on src/common/system.cpp:
```diff
@@ -113,10 +113,10 @@ int GetNumCores()
 #ifdef WIN32
     if (MEMORYSTATUSEX m{}; (m.dwLength = sizeof(m), GlobalMemoryStatusEx(&m))) return clamp(m.ullTotalPhys);
-#elif defined(__linux__) || defined(__APPLE__)
+#elif defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__)
```
Here are the relevant docs:
- [OpenBSD manpage](https://man.openbsd.org/sysconf.3)
- [NetBSD manpage](https://man.netbsd.org/sysconf.3)
- [illumos manpage](https://illumos.org/man/3C/sysconf)

l0rinc replied:
Based on https://github.com/bitcoin/bitcoin/blob/56c6daa64f6bc29fb83f3ec8940c37ddc549edeb/src/common/system.cpp#L79 I also suggest something similar:
```suggestion
#elif defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__)
```

l0rinc replied:
No, I just cannot personally test or ACK them since there's no way for me to test them, but if you do, that's enough for me.

hebasto replied:
> Does it display the correct value if we print `GetTotalRAM()`?

Yes, it does.

l0rinc replied:
`NetBSD` also finished successfully: https://github.com/l0rinc/bitcoin/actions/runs/17870748757/job/50823702246?pr=39

vasild replied:
Done.

l0rinc commented:
Thanks @hebasto for finding it, thanks @vasild for the quick PR.

I'm working on trying to reproduce this on the mentioned platforms, but there's a reason I didn't add them originally since I couldn't test them via godbolt (https://xania.org/202506/how-compiler-explorer-works indicates they don't have native BSD or s390x servers).

@vasild if you edit again, can you please add @hebasto as a coauthor for the change?

Concept ACK for both changes in the PR (the maybe unused + adding BSD), I'm trying to find a way to test it myself before fully ACKing.

l0rinc commented:
I was able to reproduce the problem with `FreeBSD` by adding a [dummy CI job](https://github.com/l0rinc/bitcoin/pull/39/commits):
* https://github.com/l0rinc/bitcoin/actions/runs/17867131621/job/50812048046?pr=39#step:3:865
> error: unused variable 'clamp'

After applying the PR the test passes successfully:
* https://github.com/l0rinc/bitcoin/actions/runs/17868501673/job/50816606849?pr=39#step:3:1394
> test/system_tests.cpp(23): info: check GetTotalRAM() >= 1000_MiB has passed
> GetTotalRAM() = 4258750464
> test/system_tests.cpp(29): info: check GetTotalRAM() < 10'000'000_MiB has passed

Which coincides with the set `memory: 4G`.

Based on the [BSD definitions](https://github.com/bitcoin/bitcoin/blob/56c6daa64f6bc29fb83f3ec8940c37ddc549edeb/src/common/system.cpp#L79) in the same file I left a suggestion but I'm also fine with merging as-is.

ACK 2b6497b03585922519a16f782be8f6c83e55ab29

hebasto reviewed (commented):
Additionally, I suggest guarding this check:https://github.com/bitcoin/bitcoin/blob/56c6daa64f6bc29fb83f3ec8940c37ddc549edeb/src/test/system_tests.cpp#L23 with the same `#ifdef`s as in the `GetTotalRAM()` implementation. Otherwise, this test will fail on any unmentioned system.

l0rinc commented:
> Otherwise, this test will fail on any unmentioned system

That's exactly why it was added so that we can detect and cover those. Is it too strict?

hebasto commented:
> > Otherwise, this test will fail on any unmentioned system
> 
> That's exactly why it was added so that we can detect and cover those. Is it too strict?

We usually avoid breaking people’s tests just to collect coverage data :)

l0rinc commented:
I assumed we want to be notified about unsupported systems, is that not the case?

l0rinc committed:
test: split out `system_ram_tests` to signal when total ram cannot be determined

when `GetTotalRAM` returns an `std::nullopt` now we're getting:
```
The following tests did not run:
        106 - system_ram_tests (Skipped)
```

l0rinc commented:
Good idea @hebasto, @vasild can you please cherry-pick [l0rinc/bitcoin@`f011c2c` (#39)](https://github.com/l0rinc/bitcoin/pull/39/commits/f011c2cdf170dbd53c7ec2e46379238579232d67) on top of the PR?
Please add @hebasto as a coauthor.

l0rinc reviewed (approved):
ACK 8fcf71ca005449e639f189dec7ec0163a07e6d37

Reproduced the unused clamp on FreeBSD and that the fix returns the correct memory on FreeBSD and NetBSD and OpenBSD as well (`__illumos__` was tested by @hebasto).
I have also tested that on systems that don't support the memory query the test is gracefully skipped and reported as skipped.

Thanks for the quick fix.

hebasto commented on src/test/system_ram_tests.cpp:
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2025-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+//
```
```suggestion
// file COPYING or https://opensource.org/license/mit/.
```

hebasto commented on src/common/system.cpp:
```diff
@@ -113,10 +113,15 @@ int GetNumCores()
 #ifdef WIN32
     if (MEMORYSTATUSEX m{}; (m.dwLength = sizeof(m), GlobalMemoryStatusEx(&m))) return clamp(m.ullTotalPhys);
-#elif defined(__linux__) || defined(__APPLE__)
+#elif defined(__APPLE__) || \
+      defined(__FreeBSD__) || \
+      defined(__NetBSD__) || \
+      defined(__OpenBSD__) || \
+      defined(__illumos__) || \
+      defined(__linux__)
```
4ebf456dc589a993faa73b8b377c0d1919fbd577

Could you please update the commit message so it reflects the changes?

vasild replied:
Done.

hebasto reviewed (approved):
ACK 8fcf71ca005449e639f189dec7ec0163a07e6d37.

hebasto reviewed (approved):
ACK 56791b582958e905e5ba5cbf172a8ea7dad1a8b0.

l0rinc commented:
ACK 56791b582958e905e5ba5cbf172a8ea7dad1a8b0
